### ISSUE TITLE
[Featured Product] Card link generates overly verbose user experience

### ISSUE THREAD

**Issue by svinkle** [Shopify/dawn#450]
## Issue summary

The featured product card links wrap the entirety of the card content:

1. Image `alt` text
2. Product title
3. Product vendor
4. Product price/sale price
5. …and other content

This generates an overly verbose user experience for screen reader and voice dictation users. Screen readers announce all content at once, or when navigating via virtual cursor, announce "link" along with each piece of content.

> "Alle Coffee Table Small by Staffan Holm Alle Coffee Table Small by Staffan Holm, Oak / Pale, Small Alle Coffee Table Small by Staffan Holm, Oak / Black, Small Vendor: HEM Alle Coffee Table Small by Staffan Holm Regular price $401.00"

Voice dictation users may not be able to call out the command to click the link with the product title; this is not the first piece of content which makes up the computed accessible name (the first piece is the image `alt` as this is first in the DOM within the link.)

<details>
  <summary>Screenshots</summary>

<img width="1153" alt="Dawn theme, featured product section cards are highlighted." src="https://user-images.githubusercontent.com/1392632/130126292-6161a6b9-40fa-4605-b4b3-d769b61c1d9a.png">

</details>

#### HTML

```html
<div class="card-wrapper">
  <a href="/products/key-side-table-by-gamfratesi" class="full-unstyled-link">
    <span class="visually-hidden">Key Side Table by GamFratesi</span>
    <div class="card card--product" tabindex="-1">
      <div class="card__inner">
        <div>
          <div
            class="media media--transparent media--adapt media--hover-effect"
            style="…"
          >
            <!-- Images… -->
          </div>
        </div>
        <div class="card__badge">
          <span
            class="badge badge--bottom-left color-accent-2"
            aria-hidden="true"
          >
            Sale
          </span>
        </div>
      </div>
    </div>
    <div class="card-information">
      <div class="card-information__wrapper">
        <!-- Vendor… -->
        <span class="card-information__text h5">
          Key Side Table by GamFratesi
        </span>
        <!-- Caption -->
        <div class="price price--on-sale">
          <!-- Price… -->
        </div>
      </div>
    </div>
  </a>
</div>
```

## Steps to reproduce

1. Go to a Dawn theme store with any supported browser
2. With an active screen reader, <kbd>Tab</kbd> to or use the virtual cursor to navigate the featured product listing

### Expected

- <kbd>Tab</kbd> to announce the product title only, virtual cursor to consume other card content with intended semantic meaning

### Actual

- <kbd>Tab</kbd> announces all content within the link, virtual cursor announces all content items with "link" semantics

### HTML structure…

1. Move the link element to only wrap the product title text.
2. Add a appropriate heading element by swapping the `span.card-information__text` element with an `h3`.
3. Remove the `span.visually-hidden` element as this only adds more verbosity to the card content.
4. Remove the `aria-hidden` element from the `span.badge` element in oder to create an equal user experience for screen reader users.
5. Reverse the DOM order of `div.card-information` and `div.card--product` in order to ensure the heading comes before the product image (otherwise screen reader users navigating by heading alone will miss the image description.)

### CSS modifications

1. Use CSS to wrap/cover the card with the link element to re-apply the large click/touch target.
2. Reverse the order of `div.card-information` and `div.card--product` using CSS in order to retain the original design.

Review [Inclusive Components – Card](https://inclusive-components.design/cards/) by @Heydon for the rationale behind these changes.

#### HTML

```html
<div class="card-wrapper">
  <div class="card-information">
    <div class="card-information__wrapper">
      <!-- Vendor… -->
      <h3 class="card-information__text h5">
        <a
          href="/products/key-side-table-by-gamfratesi"
          class="full-unstyled-link"
        >
          Key Side Table by GamFratesi
        </a>
      </h3>
      <!-- Caption -->
      <div class="price price--on-sale">
        <!-- Price… -->
      </div>
    </div>
  </div>
  <div class="card card--product" tabindex="-1">
    <div class="card__inner">
      <div>
        <div
          class="media media--transparent media--adapt media--hover-effect"
          style="…"
        >
          <!-- Images… -->
        </div>
      </div>
      <div class="card__badge">
        <span class="badge badge--bottom-left color-accent-2">Sale</span>
      </div>
    </div>
  </div>
</div>
```

#### CSS

```css
.full-unstyled-link:after {
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
}

.card-wrapper {
  display: flex;
  flex-direction: column;
  /* … */
}

.card-wrapper .card-information {
  order: 1;
}
```

<!-- #### Diff

```diff
- line removed
+ line added
``` -->

## Specifications

- **WCAG Principal:** Perceivable, Robust
- **WCAG SC:** [1.3.1 Info and Relationships](https://www.w3.org/WAI/WCAG21/Understanding/info-and-relationships.html), [4.1.2 Name, Role, Value](https://www.w3.org/WAI/WCAG21/Understanding/name-role-value.html)
- **Severity:** High
- **Effort:** High
- **Related:** #79

**Comment by svinkle** [comment]
@tauthomas01, thanks for calling this out. Yes, in this rare instance, reversing the order would be desirable. This is based on the fact that there is content "above" the `h3` heading. When screen reader users are navigating via headings alone, having all content positioned "below" the heading would allow for content consumption from a natural, top-down fashion.

This would not be a violation of [2.4.3 Focus Order](https://www.w3.org/WAI/WCAG21/Understanding/focus-order.html) due to the content in question not being focusable elements. Also, the close proximity in which the content is reversed would have little negative affect when navigating via virtual/browse mode.
